// react-instantsearch-dom/dist/cjs/core/createInstantSearchServer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _isEmpty2 = require('lodash/isEmpty');

var _isEmpty3 = _interopRequireDefault(_isEmpty2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _server = require('react-dom/server');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _algoliasearchHelper = require('algoliasearch-helper');

var _algoliasearchHelper2 = _interopRequireDefault(_algoliasearchHelper);

var _reactInstantsearchCore = require('react-instantsearch-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var getIndex = function getIndex(context) {
  return context && context.multiIndexContext ? context.multiIndexContext.targetedIndex : context.ais.mainTargetedIndex;
};

var hasMultipleIndex = function hasMultipleIndex(context) {
  return context && context.multiIndexContext;
};

var createInstantSearchServer = function createInstantSearchServer(algoliasearch) {
  var InstantSearch = (0, _reactInstantsearchCore.createInstantSearch)(algoliasearch, {
    Root: 'div',
    props: { className: 'ais-InstantSearch__root' }
  });

  var searchParameters = [];
  var client = void 0;
  var indexName = '';

  var onSearchParameters = function onSearchParameters(getSearchParameters, context, props, searchState) {
    var index = getIndex(context);
    searchParameters.push({
      getSearchParameters: getSearchParameters,
      context: context,
      props: props,
      searchState: searchState,
      index: index
    });
  };

  var findResultsState = function findResultsState(App, props) {
    searchParameters = [];

    (0, _server.renderToString)(_react2.default.createElement(App, props));

    var sharedSearchParameters = searchParameters.filter(function (searchParameter) {
      return !hasMultipleIndex(searchParameter.context) && (searchParameter.props.indexName === indexName || !searchParameter.props.indexName);
    }).reduce(function (acc, searchParameter) {
      return searchParameter.getSearchParameters.call({ context: searchParameter.context }, acc, searchParameter.props, searchParameter.searchState);
    }, new _algoliasearchHelper.SearchParameters(_extends({ index: indexName }, _reactInstantsearchCore.HIGHLIGHT_TAGS)));

    var mergedSearchParameters = searchParameters.filter(function (searchParameter) {
      return hasMultipleIndex(searchParameter.context);
    }).reduce(function (acc, searchParameter) {
      var index = getIndex(searchParameter.context);
      var sp = searchParameter.getSearchParameters.call({ context: searchParameter.context }, acc[index] ? acc[index] : sharedSearchParameters, searchParameter.props, searchParameter.searchState);
      acc[index] = sp;
      return acc;
    }, {});

    if ((0, _isEmpty3.default)(mergedSearchParameters)) {
      var helper = (0, _algoliasearchHelper2.default)(client, sharedSearchParameters.index);
      return helper.searchOnce(sharedSearchParameters);
    } else {
      var _helper = (0, _algoliasearchHelper2.default)(client, sharedSearchParameters.index);
      var search = [];

      if (mergedSearchParameters[indexName]) {
        search.push(_helper.searchOnce(_extends({}, sharedSearchParameters, mergedSearchParameters[sharedSearchParameters.index], {
          index: mergedSearchParameters[sharedSearchParameters.index].index
        })));
        delete mergedSearchParameters[indexName];
      } else {
        search.push(_helper.searchOnce(sharedSearchParameters));
      }

      search.push.apply(search, _toConsumableArray(Object.keys(mergedSearchParameters).map(function (key) {
        var derivedHelper = (0, _algoliasearchHelper2.default)(client, mergedSearchParameters[key].index);
        return derivedHelper.searchOnce(mergedSearchParameters[key]);
      })));

      return Promise.all(search);
    }
  };

  var decorateResults = function decorateResults(results) {
    if (!results) {
      return undefined;
    }

    return Array.isArray(results) ? results.reduce(function (acc, result) {
      acc[result.state.index] = new _algoliasearchHelper.SearchResults(new _algoliasearchHelper.SearchParameters(result.state), result._originalResponse.results);
      return acc;
    }, []) : new _algoliasearchHelper.SearchResults(new _algoliasearchHelper.SearchParameters(results.state), results._originalResponse.results);
  };

  var CreateInstantSearchServer = function (_Component) {
    _inherits(CreateInstantSearchServer, _Component);

    function CreateInstantSearchServer(props) {
      _classCallCheck(this, CreateInstantSearchServer);

      var _this = _possibleConstructorReturn(this, (CreateInstantSearchServer.__proto__ || Object.getPrototypeOf(CreateInstantSearchServer)).call(this));

      if (props.searchClient) {
        if (props.appId || props.apiKey || props.algoliaClient) {
          throw new Error('react-instantsearch:: `searchClient` cannot be used with `appId`, `apiKey` or `algoliaClient`.');
        }
      }

      if (props.algoliaClient) {
        // eslint-disable-next-line no-console
        console.warn('`algoliaClient` option was renamed `searchClient`. Please use this new option before the next major version.');
      }

      client = props.searchClient || props.algoliaClient || algoliasearch(props.appId, props.apiKey);

      if (typeof client.addAlgoliaAgent === 'function') {
        client.addAlgoliaAgent('react-instantsearch ' + _reactInstantsearchCore.version);
      }

      indexName = props.indexName;
      return _this;
    }

    _createClass(CreateInstantSearchServer, [{
      key: 'render',
      value: function render() {
        var resultsState = decorateResults(this.props.resultsState);
        return _react2.default.createElement(InstantSearch, _extends({}, this.props, {
          resultsState: resultsState,
          onSearchParameters: onSearchParameters
        }));
      }
    }]);

    return CreateInstantSearchServer;
  }(_react.Component);

  CreateInstantSearchServer.propTypes = {
    algoliaClient: _propTypes2.default.object,
    searchClient: _propTypes2.default.object,
    appId: _propTypes2.default.string,
    apiKey: _propTypes2.default.string,
    indexName: _propTypes2.default.string.isRequired,
    resultsState: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.array])
  };


  return { InstantSearch: CreateInstantSearchServer, findResultsState: findResultsState };
};

exports.default = createInstantSearchServer;
