// react-instantsearch-core/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createInstantSearch = require('./core/createInstantSearch');

Object.defineProperty(exports, 'createInstantSearch', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_createInstantSearch).default;
  }
});

var _createIndex = require('./core/createIndex');

Object.defineProperty(exports, 'createIndex', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_createIndex).default;
  }
});

var _createConnector = require('./core/createConnector');

Object.defineProperty(exports, 'createConnector', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_createConnector).default;
  }
});

var _highlight = require('./core/highlight');

Object.defineProperty(exports, 'HIGHLIGHT_TAGS', {
  enumerable: true,
  get: function get() {
    return _highlight.HIGHLIGHT_TAGS;
  }
});

var _version = require('./core/version');

Object.defineProperty(exports, 'version', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_version).default;
  }
});

var _Configure = require('./widgets/Configure');

Object.defineProperty(exports, 'Configure', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_Configure).default;
  }
});

var _connectAutoComplete = require('./connectors/connectAutoComplete');

Object.defineProperty(exports, 'connectAutoComplete', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectAutoComplete).default;
  }
});

var _connectBreadcrumb = require('./connectors/connectBreadcrumb');

Object.defineProperty(exports, 'connectBreadcrumb', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectBreadcrumb).default;
  }
});

var _connectConfigure = require('./connectors/connectConfigure');

Object.defineProperty(exports, 'connectConfigure', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectConfigure).default;
  }
});

var _connectCurrentRefinements = require('./connectors/connectCurrentRefinements');

Object.defineProperty(exports, 'connectCurrentRefinements', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectCurrentRefinements).default;
  }
});

var _connectHierarchicalMenu = require('./connectors/connectHierarchicalMenu');

Object.defineProperty(exports, 'connectHierarchicalMenu', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectHierarchicalMenu).default;
  }
});

var _connectHighlight = require('./connectors/connectHighlight');

Object.defineProperty(exports, 'connectHighlight', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectHighlight).default;
  }
});

var _connectHits = require('./connectors/connectHits');

Object.defineProperty(exports, 'connectHits', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectHits).default;
  }
});

var _connectHitsPerPage = require('./connectors/connectHitsPerPage');

Object.defineProperty(exports, 'connectHitsPerPage', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectHitsPerPage).default;
  }
});

var _connectInfiniteHits = require('./connectors/connectInfiniteHits');

Object.defineProperty(exports, 'connectInfiniteHits', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectInfiniteHits).default;
  }
});

var _connectMenu = require('./connectors/connectMenu');

Object.defineProperty(exports, 'connectMenu', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectMenu).default;
  }
});

var _connectNumericMenu = require('./connectors/connectNumericMenu');

Object.defineProperty(exports, 'connectNumericMenu', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectNumericMenu).default;
  }
});

var _connectPagination = require('./connectors/connectPagination');

Object.defineProperty(exports, 'connectPagination', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectPagination).default;
  }
});

var _connectPoweredBy = require('./connectors/connectPoweredBy');

Object.defineProperty(exports, 'connectPoweredBy', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectPoweredBy).default;
  }
});

var _connectRange = require('./connectors/connectRange');

Object.defineProperty(exports, 'connectRange', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectRange).default;
  }
});

var _connectRefinementList = require('./connectors/connectRefinementList');

Object.defineProperty(exports, 'connectRefinementList', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectRefinementList).default;
  }
});

var _connectScrollTo = require('./connectors/connectScrollTo');

Object.defineProperty(exports, 'connectScrollTo', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectScrollTo).default;
  }
});

var _connectSearchBox = require('./connectors/connectSearchBox');

Object.defineProperty(exports, 'connectSearchBox', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectSearchBox).default;
  }
});

var _connectSortBy = require('./connectors/connectSortBy');

Object.defineProperty(exports, 'connectSortBy', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectSortBy).default;
  }
});

var _connectStateResults = require('./connectors/connectStateResults');

Object.defineProperty(exports, 'connectStateResults', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectStateResults).default;
  }
});

var _connectStats = require('./connectors/connectStats');

Object.defineProperty(exports, 'connectStats', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectStats).default;
  }
});

var _connectToggleRefinement = require('./connectors/connectToggleRefinement');

Object.defineProperty(exports, 'connectToggleRefinement', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectToggleRefinement).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
